stackslib/src/net/api/postblock_proposal.rs

```diff
@@ -0,0 +1,220 @@
+        serde_json::from_slice(body).map_err(|_| Error::DecodeError("Failed to parse body".into()))
+    }
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCBlockProposalRequestHandler {
+    fn verb(&self) -> &'static str {
+        "POST"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(r#"^/v2/block_proposal$"#).unwrap()
+    }
+
+    /// Try to decode this request.
+    /// There's nothing to load here, so just make sure the request is well-formed.
+    fn try_parse_request(
+        &mut self,
+        preamble: &HttpRequestPreamble,
+        _captures: &Captures,
+        query: Option<&str>,
+        body: &[u8],
+    ) -> Result<HttpRequestContents, Error> {
+        // Only accept requests from localhost
+        let is_loopback = match preamble.host {
+            // Should never be DNS
+            PeerHost::DNS(..) => false,
+            PeerHost::IP(addr, ..) => addr.is_loopback(),
+        };
+
+        if !is_loopback {
+            return Err(Error::Http(403, "Forbidden".into()));
+        }
+
+        if preamble.get_content_length() == 0 {
+            return Err(Error::DecodeError(
+                "Invalid Http request: expected non-zero-length body for PostBlock".to_string(),
+            ));
+        }
+
+        if preamble.get_content_length() > MAX_PAYLOAD_LEN {
+            return Err(Error::DecodeError(
+                "Invalid Http request: BlockProposal body is too big".to_string(),
+            ));
+        }
+
+        let block_proposal = match preamble.content_type {
+            Some(HttpContentType::Bytes) => Self::parse_octets(body)?,
+            Some(HttpContentType::JSON) => Self::parse_json(body)?,
+            None => {
+                return Err(Error::DecodeError(
+                    "Missing Content-Type for transaction".to_string(),
+                ))
+            }
+            _ => {
+                return Err(Error::DecodeError(
+                    "Wrong Content-Type for transaction; expected application/json or application/octet-stream".to_string(),
+                ))
+            }
+        };
+
+        self.block_proposal = Some(block_proposal);
+        Ok(HttpRequestContents::new().query_string(query))
+    }
+}
+
+impl RPCRequestHandler for RPCBlockProposalRequestHandler {
+    /// Reset internal state
+    fn restart(&mut self) {
+        self.block_proposal = None
+    }
+
+    /// Make the response
+    fn try_handle_request(
+        &mut self,
+        preamble: HttpRequestPreamble,
+        _contents: HttpRequestContents,
+        node: &mut StacksNodeState,
+    ) -> Result<(HttpResponsePreamble, HttpResponseContents), NetError> {
+        let block_proposal = self
+            .block_proposal
+            .take()
+            .ok_or(NetError::SendError("`block_proposal` not set".into()))?;
+
+        let res = node.with_node_state(|_network, sortdb, chainstate, _mempool, _rpc_args| {
```

Comment:
You may need to modify this function `with_node_state()` to pass `&mut SortitionDB`.  Alternatively, you can add a new function `with_node_state_mut()` which passes `&mut` references for all of these things.

Reply:
You may need to tweak `PeerNetwork::run()` to take a `&mut SortitionDB` as well

Reply:
Actually, forget I said this.  Let's try something else.